src/app/data-explorer/data-specification.service.ts

```diff
@@ -542,34 +541,44 @@ export class DataSpecificationService {
   ): Observable<DataSpecification> {
     return forkJoin([
       // TODO: assume there is only one data model, will have to change in future
+      // Get the rootModel and create a new empty data specification.
       this.dataExplorer.getRootDataModel(),
       this.create(user, name, description),
     ]).pipe(
       switchMap(([rootDataModel, dataSpecification]) => {
-        if (!rootDataModel || !rootDataModel.id) {
-          return throwError(() => new Error('No root data model'));
+        if (!rootDataModel?.id || !dataSpecification?.id) {
+          return throwError(() => new Error('Missing root data model or data specification'));
         }
 
-        if (!dataSpecification || !dataSpecification.id) {
-          return throwError(() => new Error('No data specification'));
-        }
-
-        return this.copyDataElementsToDataModel(
-          rootDataModel,
+        // Get the result of copying the elements to the new data specification.
+        const populatedDataSpecification$ = this.dataModels.copySubset(
           rootDataModel.id,
-          elements,
-          dataSpecification.id
+          dataSpecification.id,
+          {
+            additions: elements.map((de) => de.id),
+            deletions: [],
+          }
         );
+
+        // Get the coreTableProfile.
+        const coreTableProfile$ =
+          this.coreTableProfileService.getQueryBuilderCoreTableProfile(rootDataModel);
+
+        return forkJoin({
+          populatedDataSpecification: populatedDataSpecification$,
+          coreTableProfile: coreTableProfile$,
+        });
       }),
-      switchMap(([rootDataModel, dataSpecification]) => {
-        return this.getCoreTableProfile(rootDataModel, dataSpecification);
-      }),
-      switchMap(([dataSpecification, coreTableProfile]) => {
-        return this.saveCoreTableProfile(dataSpecification, coreTableProfile);
-      }),
-      map(([_, targetDataModel]) =>
-        this.submissionSDEService.mapToDataSpecification(targetDataModel)
-      )
+      map(({ populatedDataSpecification, coreTableProfile }) => {
+        // Save the core table profile to the newly created data specification.
+        this.coreTableProfileService.saveQueryBuilderCoreTableProfile(
+          populatedDataSpecification,
+          coreTableProfile
+        );
+
+        // We don't care about the output above so just return the data specification.
+        return this.submissionSDEService.mapToDataSpecification(populatedDataSpecification);
+      })
```

Comment:
This is not working? When I try and create a specification and go to look at the query builder I get the error:
```
"Mauro Data Explorer - Query Builder Core Table" (data model profile) for DataModel validation errors:
- This field cannot be empty
- Format needs to be [Schema].[Table] e.g: people.patient
```
I checked the network log and I can see the profile being "got", but I cannot see it being saved.

Reply:
This is due to the fact that I was incorrectly using `map` here. The fix is to use `switchMap` to switch to the observable emitted by the `saveQueryBuilderCoreTableProfile()` method and then pipe off the end of that to map the dataSpecification properly. This ensures the save call completes successfully before moving along to the next step.